### ISSUE TITLE
Webstarted application tries to load resources from server incorrectly

### ISSUE THREAD

**Issue by AndreasEhret** [AdoptOpenJDK/IcedTea-Web#89]
This issue was imported from bugzilla with bugzilla2github
# Bugzilla Bug Referenze: [3705](https://icedtea.classpath.org/bugzilla/show_bug.cgi?id=3705)
Date: 2019-02-21 12:47:00 +0000
From: Christian Svedin &lt;<christian.svedin>&gt;
To:   JiriVanek &lt;<jvanek>&gt;
CC:   rohit.bisht.3612, unassigned

Last updated: 2019-03-14 09:04:16 +0000

## Comment 17417
Date: 2019-02-21 12:47:58 +0000
From: Christian Svedin &lt;<christian.svedin>&gt;

According to this (
                https://docs.oracle.com/javase/8/docs/technotes/guides/javaws/developersguide/faq.html#s113 ) webstart
                is not supposed to try to download all kinds of resources from the server, only jars, but icedtea-web
                has a classloader that has a fallback to always download resources from the server, no matter what they
                are.

                This has performance implications. For example when using ResourceBundle it tries to get the bundle
                using different names to find the most specific one, and each try results in a new server call.

                Text from linked page:
                > Can I deploy any application with Java Web Start?
                >
                > Java Web Start is an application launcher for Java applications that are written to be
                web-deployed.
                >
                > An application must be delivered in a set of JAR files and all application resources, such as
                images, configuration files, and native libraries must be included in the JAR files. The resources must
                be looked up using the method ClassLoader.getResource() or an equivalent method. Java Web Start only
                transfers JAR files from the web server to the client.
                >
                > If an application is written to run in a restricted execution environment (sandbox), then access to
                disk is not permitted and the application may only connect to the host on which it resides.
            

## Comment 17418
Date: 2019-02-21 12:56:40 +0000
From: JiriVanek &lt;<jvanek>&gt;

Please if you can,. add reproducer.

## Comment 17419
Date: 2019-02-21 14:21:05 +0000
From: Christian Svedin &lt;<christian.svedin>&gt;

(In reply to JiriVanek from comment #1)
                > Please if you can,. add reproducer.

                I don't have a problem doing that, but I'm not sure how such a reproducer should look to help you out.
                You need a main class that does getResource() of some resource that's not available, and some way to
                check if the server is called.
                So, do you just want a .jnlp and a .jar ? or something more to handle the server side as well? Maybe you
                have an example of a reproducer for something else that both client and server and I can do something
                similar.
            

## Comment 17456
Date: 2019-03-14 05:48:39 +0000
From: Rohit Bisht &lt;<rohit.bisht.3612>&gt;

The problem here is that it is adding all the paths to codebase loader even when we should not.
                Apparently there is a boolean variable (enableCodeBase) which is used to check whether to add resource
                to the codebase loader or not.
                So just putting a check for enableCodeBase can resolve the issue.

                So for my issue all I needed was to replace

                addToCodeBaseLoader(extLoader.file.getCodeBase());

                with

                if(this.enableCodeBase) {
                addToCodeBaseLoader(extLoader.file.getCodeBase());
                }

                I guess the developers created the variable to be checked but forgot to do so or maybe had some other
                reasons not to use it.
            

## Comment 17457
Date: 2019-03-14 05:50:57 +0000
From: Rohit Bisht &lt;<rohit.bisht.3612>&gt;

There is a method enableCodeBase() where this variable can/should be checked.

## Comment 17458
Date: 2019-03-14 09:04:16 +0000
From: JiriVanek &lt;<jvanek>&gt;

hi Rohit!

                thank you very much for eyball. sounds reasonable!

**Comment by bourgesl** [comment]
I wonder if this one-liner fix could be backported to iced-tea-web 1.8 branch to make this performance fix available in the coming 1.8.x release (for most linux users) as 2.0.x is still not publicly released (alpha13 - 2020.4).

**Comment by douph1** [comment]
@bourgesl you can take a look at the bottom of https://github.com/AdoptOpenJDK/IcedTea-Web/actions/runs/621216354 to download/test the next release with this fix

**Comment by bourgesl** [comment]
Amazing,
Thanks for this early-build of Iced tea web 1.8.7.

As I currently use my own build of javaws.jar over the 1.8.3 portable build, I will test yours.

**Comment by douph1** [comment]
@bourgesl, have you test the build from here ? https://github.com/AdoptOpenJDK/IcedTea-Web/actions/runs/621216354
I waiting you to test to release the 1.8.7

**Comment by bourgesl** [comment]
Sorry, I forgot to answer.
I actually installed the icedtea-web_build_x86-64_portable 1.8.7 release since March 4th and it works like a charm.
I tested following JDK (JAVA_HOME set):
- OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_202-b08): OK
- OpenJDK Runtime Environment (Zulu 8.52.0.23-CA-linux64) (build 1.8.0_282-b08): OK
- OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.6+10)
- OpenJDK Runtime Environment AdoptOpenJDK (build 12+33): OK

With OpenJDK 13 to 17, it fails with the following exception:
```
Exception in thread "main" java.lang.NoClassDefFoundError: sun/awt/AWTSecurityManager
	at java.desktop/net.sourceforge.jnlp.runtime.JNLPRuntime.initialize(JNLPRuntime.java:262)
	at java.desktop/net.sourceforge.jnlp.runtime.Boot.init(Boot.java:353)
	at java.desktop/net.sourceforge.jnlp.runtime.JnlpBoot.run(JnlpBoot.java:58)
	at java.desktop/net.sourceforge.jnlp.runtime.Boot.run(Boot.java:274)
	at java.desktop/net.sourceforge.jnlp.runtime.Boot.run(Boot.java:63)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:312)
	at java.desktop/net.sourceforge.jnlp.runtime.Boot.main(Boot.java:214)
```
I suppose this class has been removed or made inaccessible (jigsaw encapsulation ?) since OpenJDK 13.

In conclusion: 
this release works well on OpenJDK8 & 11 (LTS), but it does not work on OpenJDK16 (current) or coming OpenJDK 17 (LTS), but it is another compatibility problem.

Thanks

**Comment by bourgesl** [comment]
Finally, I managed patching icedtea-web 1.8 branch using PR #438 (remove AWTSecurityManager use) and this patched (again) variant works now on OpenJDK 15 to 17 (not tested on previous versions).

I would recommend applying this patch on the 1.8 branch too.
